### PR TITLE
implement MaybeEndSync

### PR DESCRIPTION
**Pull request by phritz**

### CODE REVIEW

**Comment by phritz** [comment]
@aboodman can you look at how I did Mutation.args? Will that work? If so why didn't I just do that for patch value instead of stringifying it? Sorry I lost the thread on that.

File: src/sync/mod.rs [hunk 1]
```diff
@@ -139,6 +148,114 @@ pub enum BeginSyncError {
...
+) -> Result<MaybeEndSyncResponse, MaybeEndSyncError> {
+    use MaybeEndSyncError::*;
+
+    // Ensure sync head is what the caller thinks it is.
+    let mut dag_write = store.write().await.map_err(WriteError)?;
+    let dag_read = dag_write.read();
+    let sync_head_hash = dag_read
+        .get_head(SYNC_HEAD_NAME)
+        .await
+        .map_err(ReadError)?
```
Comment (aboodman) [inline src/sync/mod.rs:163]: `GetSyncHeadError`?

File: src/sync/mod.rs [hunk 2]
```diff
@@ -139,6 +148,114 @@ pub enum BeginSyncError {
...
+    // Return replay commits if any.
+    if !pending.is_empty() {
+        let mut replay_mutations: Vec<Mutation> = Vec::new();
+        for c in pending {
+            let (name, args) = match c.meta().typed() {
+                MetaTyped::Local(lm) => (
+                    lm.mutator_name().to_string(),
+                    lm.mutator_args_json().to_vec(),
+                ),
+                _ => return Err(ProgrammerError("pending mutation is not local".to_string())),
```
Comment (aboodman) [inline src/sync/mod.rs:209]: Can we change the name `ProgrammerError` to `InternalError` throughout? Our customers are also programmers and might not understand the error is something we did wrong, not them.

File: src/sync/mod.rs [hunk 3]
```diff
@@ -139,6 +148,114 @@ pub enum BeginSyncError {
...
+        });
+    }
+
+    // TODO check invariants
+
+    // No mutations to replay so set the main head to the sync head and sync complete!
+    dag_write
+        .set_head(db::DEFAULT_HEAD_NAME, Some(&sync_head_hash))
+        .await
+        .map_err(WriteError)?;
```
Comment (aboodman) [inline src/sync/mod.rs:230]: These cases should have their own error tags so we can tell them apart if/when they happen.

File: src/sync/mod.rs [hunk 4]
```diff
@@ -139,6 +148,114 @@ pub enum BeginSyncError {
...
+    MissingMainHead,
+    MissingSyncHead,
+    NoBaseSnapshot(db::BaseSnapshotError),
+    OverlappingSyncs,
+    PendingError(db::PendingError),
+    ProgrammerError(String),
+    ReadError(dag::Error),
+    SyncSnapshotWithNoBasis,
+    WriteError(dag::Error),
+    WrongSyncHead,
```
Comment (aboodman) [inline src/sync/mod.rs:256]: OK there are admittedly a lot of cases here but this is maybe the most complex function in the entire system and I'd argue it's useful to understand the difference between all these different cases.

**Comment by aboodman** [comment]
This is really nice work @phritz .

File: src/db/commit.rs [hunk 5]
```diff
@@ -210,6 +210,30 @@ impl Commit {
         Ok(commit)
     }
 
+    // Vector returned in chain-head-first order.
+    pub async fn pending(
```
[the remark this answers is not in the document]
Comment (phritz) [reply src/db/commit.rs:214]: clarified in https://github.com/rocicorp/repc/pull/193

File: src/db/commit.rs [hunk 6]
```diff
@@ -210,6 +210,30 @@ impl Commit {
...
+        dag_read: &dag::Read<'_>,
+    ) -> Result<Vec<Commit>, PendingError> {
+        use PendingError::*;
+        let mut commit = Commit::from_hash(hash, dag_read)
+            .await
+            .map_err(NoSuchCommit)?;
+        let mut commits = Vec::new();
+        while !commit.meta().is_snapshot() {
+            let meta = commit.meta();
+            let basis_hash = meta
```
[the remark this answers is not in the document]
Comment (phritz) [reply src/db/commit.rs:225]: The str returned from basis_hash is borrowed from the commit's fb via meta, so we can't hold a reference to it while we move commit into the commits vector. to_string() enables the commit to be moved.

File: src/sync/mod.rs [hunk 7]
```diff
@@ -18,6 +18,15 @@ use std::str::FromStr;
 
 pub const SYNC_HEAD_NAME: &str = "sync";
 
+#[derive(Debug, DeJson, SerJson)]
+pub struct Mutation {
+    id: u64,
+    name: String,
+    args: Vec<u8>,
+    #[nserde(skip_serializing_if = "Option::is_none")]
```
[the remark this answers is not in the document]
Comment (phritz) [reply src/sync/mod.rs:26]: It serializes it as null: https://serde.rs/impl-serialize.html#other-special-cases. Seems like we might need to go check all of the other Options in the code and add skip_serializing_if?

File: src/sync/mod.rs [hunk 8]
```diff
@@ -139,6 +148,114 @@ pub enum BeginSyncError {
...
+    }
+
+    // Ensure another sync has not landed a new snapshot on the main chain.
+    let sync_snapshot = Commit::base_snapshot(&sync_head_hash, &dag_read)
+        .await
+        .map_err(NoBaseSnapshot)?;
+    let main_head_hash = dag_read
+        .get_head(db::DEFAULT_HEAD_NAME)
+        .await
+        .map_err(ReadError)?
```
[the remark this answers is not in the document]
Comment (phritz) [reply src/sync/mod.rs:176]: done in https://github.com/rocicorp/repc/pull/193

File: src/sync/mod.rs (hunk 1 above)
Comment (phritz) [reply src/sync/mod.rs:163]: done in https://github.com/rocicorp/repc/pull/193

File: src/sync/mod.rs (hunk 2 above)
Comment (phritz) [reply src/sync/mod.rs:209]: done in https://github.com/rocicorp/repc/pull/193

File: src/sync/mod.rs (hunk 3 above)
Comment (phritz) [reply src/sync/mod.rs:230]: done in https://github.com/rocicorp/repc/pull/193

File: src/sync/mod.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (aboodman) [reply src/sync/mod.rs:26]: No strong opinion.

**Pull request merged**